src/components/SearchBox/SearchBox.js

```diff
@@ -0,0 +1,213 @@
+    templates: PropTypes.object.isRequired,
+    query: PropTypes.string,
+    showSubmit: PropTypes.bool,
+    showReset: PropTypes.bool,
+    showLoadingIndicator: PropTypes.bool,
+    refine: PropTypes.func,
+    autofocus: PropTypes.bool,
+    searchAsYouType: PropTypes.bool,
+    isSearchStalled: PropTypes.bool,
+    disabled: PropTypes.bool,
+    onChange: PropTypes.func,
+    onSubmit: PropTypes.func,
+    onReset: PropTypes.func,
+  };
+
+  static defaultProps = {
+    query: '',
+    showSubmit: true,
+    showReset: true,
+    showLoadingIndicator: true,
+    autofocus: false,
+    searchAsYouType: true,
+    isSearchStalled: false,
+    disabled: false,
+    onChange: noop,
+    onSubmit: noop,
+    onReset: noop,
+    refine: noop,
+  };
+
+  state = {
+    query: this.props.searchAsYouType ? '' : this.props.query,
+  };
+
+  componentWillReceiveProps(nextProps) {
+    // This keeps track of the `searchParameters` and changes the query accordingly.
+    if (!nextProps.searchAsYouType && nextProps.query !== this.props.query) {
+      this.setState({
+        query: nextProps.query,
+      });
+    }
+  }
+
+  /**
+   * This public method is used in the RefinementList SFFV search box
+   * to reset the input state when an item is selected.
+   *
+   * @see RefinementList#componentWillReceiveProps
+   * @return {undefined}
+   */
+  resetInput() {
+    this.setState({ query: '' });
+  }
+
+  onInputRef = input => {
+    this.input = input;
+
+    if (this.props.autofocus) {
+      this.input.focus();
+    }
+  };
+
+  onChange = event => {
+    const { searchAsYouType, refine, onChange } = this.props;
+    const query = event.target.value;
+
+    if (searchAsYouType) {
+      refine(query);
+    } else {
+      this.setState({ query });
+    }
+
+    onChange(event);
+  };
+
+  onSubmit = event => {
+    const { searchAsYouType, refine, onSubmit } = this.props;
+
+    event.preventDefault();
+    event.stopPropagation();
+    this.input.blur();
+
+    if (!searchAsYouType) {
+      refine(this.state.query);
+    }
+
+    onSubmit(event);
+
+    return false;
+  };
+
+  onReset = event => {
+    const { searchAsYouType, refine, onReset } = this.props;
+    const query = '';
+
+    this.input.focus();
+
+    refine(query);
+
+    if (!searchAsYouType) {
+      this.setState({ query });
+    }
+
+    onReset(event);
+  };
+
+  render() {
+    const {
+      cssClasses,
+      placeholder,
+      autofocus,
+      showSubmit,
+      showReset,
+      showLoadingIndicator,
+      templates,
+      isSearchStalled,
+      searchAsYouType,
+    } = this.props;
+
+    const query = searchAsYouType ? this.props.query : this.state.query;
+
+    return (
+      <div className={cssClasses.root}>
+        <form
+          action=""
+          role="search"
+          className={cssClasses.form}
+          noValidate
+          onSubmit={this.onSubmit}
+          onReset={this.onReset}
+        >
+          <input
+            ref={this.onInputRef}
```

Comment:
I haven't seen the `onInputRef` pattern before. Since we already pass `autoFocus` it seems redundant

Reply:
Right, I'll simplify this.

I thought it was for compatibility reasons that we force the focus with `focus()` even though we use the `autofocus` attribute.